### ISSUE TITLE
`ivadomed` input dimensions change of order compared to actual input dimensions (via `nibabel`)

### ISSUE THREAD

**Issue by uzaymacar** [ivadomed/model_seg_ms_mp2rage#26]
~~**This is potentially a bug that will be transferred to the main `ivadomed` repository (as an issue and/or PR) after verification.**~~

The input dimensions of subject images read via `nibabel` are in a different order than the `NumPy` arrays inside `ivadomed/transforms.py`. I haven't done a full trace yet, but it is possible that during processing / loading of the data the order of axes are somehow altered with. Another alternative is that the subject images we are working with are somehow problematic.

To reproduce this error, first (i) clone this repository, (ii) clone the data, and (iii) run preprocessing as shown in the main comment of #11. Whereas `basel-mp2rage-preprocessed/data_processed_lesionseg/sub-P005/anat/sub-P005_UNIT1.nii.gz` has a shape of `(42, 105, 28)` as read via `nibabel`, this will be converted to `(28, 42, 105)` in `ivadomed`. This can be verified with adding a `print(sample.shape)` inside `ivadomed/transforms.py`, for example [here](https://github.com/ivadomed/ivadomed/blob/d1b9e118f491d9e5e88523f8d0f0cae65b26ecca/ivadomed/transforms.py#L464).

Currently, we address this change of order in #11 by changing the order of the center crop, length, and stride parameters accordingly as mentioned [here](https://github.com/ivadomed/model_seg_ms_mp2rage/pull/11#issuecomment-998691726). The validity of this approach (and hence the suspicion for a bug) is supported by the immensely improved results shown in this week's [progress report](https://docs.google.com/presentation/d/1xgR1ZBwQx3jXfWUqWtTvixTIwCiaohkpkm2vCwTxJlw/edit#slide=id.g107e0779dea_0_0).

(Another possibility is that this expected behavior, and we can simply reshape the output --> to verify)

**Comment by jcohenadad** [comment]
I suspect this issue is related to the improper usage of the [`slice_axis`](https://ivadomed.org/configuration_file.html?highlight=orientation#slice-axis) field in `ivadomed`.

First, let's look at the orientation of the input image:
~~~
data_basel-mp2rage_scseg_20211217_104003/data_processed_clean/sub-P001/anat $ fslhd sub-P001_UNIT1

qform_xorient	Anterior-to-Posterior
qform_yorient	Superior-to-Inferior
qform_zorient	Left-to-Right
~~~

--> The input image has the 3rd dimension oriented along L-R, meaning that the orientation is sagittal.

Next, let's look at the [config file used](https://github.com/ivadomed/model_seg_ms_mp2rage/blob/cb6708c72e26b39b9de61f72f5fd981b9e889392/config/seg_lesion.json). Especially, the field `slice_axis`:

https://github.com/ivadomed/model_seg_ms_mp2rage/blob/cb6708c72e26b39b9de61f72f5fd981b9e889392/config/seg_lesion.json#L33

Here, it says "axial", meaning that ivadomed is expecting an axial image, whereas in fact the orientation is sagittal.

This has likely caused the confusion about the cropping size along each dimension to produce unexpected results:

https://github.com/ivadomed/model_seg_ms_mp2rage/blob/cb6708c72e26b39b9de61f72f5fd981b9e889392/config/seg_lesion.json#L97

Given that other users might fall in the same trap, we might want to think of alternative solutions to make it less confusing for our users.

My suggestion for now is to change `slice_axis` for `sagittal`

**Comment by uzaymacar** [comment]
@jcohenadad There is still something that I can't comprehend about this. Taking `basel-mp2rage-preprocessed/data_processed_lesionseg/sub-P005/anat/sub-P005_UNIT1.nii.gz` as the current example, below is table summarizing the shapes we get with different load methods:

| Load Method | Shape |
| ------------- | ------ |
| `nibabel` | `(42, 105, 28)` |
| `ivadomed`, `"slice_axis": "axial"` | `(28, 42, 105)` |
| `ivadomed`, `"slice_axis": "sagittal"` | `(105, 42, 28)` |
| `ivadomed`, `"slice_axis": "coronal"` | `(105, 28, 42)` |

`nibabel` doesn't match with any of the shapes. I am wondering what is "correct" here. For now, I will take `ivadomed`, `"slice_axis": "sagittal"` (due to image orientation) as reference and adapt the parameters accordingly.

**Comment by jcohenadad** [comment]
> nibabel doesn't match with any of the shapes. I am wondering what is "correct" here. For now, I will take ivadomed, "slice_axis": "sagittal" (due to image orientation) as reference and adapt the parameters accordingly.

yup, i am not surprised there could be inconsistencies. Ivadomed is only interested in the _last_ dimension (ie: slice), so it did what you asked for: nibabel's 3rd dim is "28" and ivadomed's 3rd dim is "28". However, the X and Y dims are swapped, but that could happen as ivadomed was not designed to _also_ account for these. 

In order to solve that issue, we would need to also check for the 1st and 2nd dimensions, and instead of only asking for the slice, we should ask for the orientation, e.g. RPI, LAS, etc.

Discussion with the ivadomed's team is required here.

**Comment by uzaymacar** [comment]
> so it did what you asked for: nibabel's 3rd dim is "28" and ivadomed's 3rd dim is "28"

I see, thanks for the explanation! I will stick to `sagittal` and use `(128, 64, 32)` for center-crop, length, and stride params.

> Discussion with the ivadomed's team is required here.

Agreed! Just to be more clear why I had an issue with this (and to correct any misconceptions I have if there are any) in the first place: When I work on a new task, I check the resolutions and shapes of the images (e.g. as seen with [this QC script](https://github.com/ivadomed/model_seg_ms_mp2rage/blob/main/preprocessing/qc_preprocess.py) in this project) via `nibabel`. When `nibabel` gives me the shape as `(X, Y, Z)`, I design center-crop, length, and stride parameters with this shape in mind. 

Now I learn that this will NOT work in `ivadomed`. I wonder what the correct approach here is. I get [this table](https://github.com/ivadomed/model_seg_ms_mp2rage/issues/26#issuecomment-999653009) by inserting a `print` line inside `ivadomed`, but obviously that's not ideal.

**Comment by uzaymacar** [comment]
Closing this issue as 
* merged PRs #11 and #19 implement the discussed changes on `"slice_axis"` key in the ivadomed config files
* an ivadomed issue [#1023](https://github.com/ivadomed/ivadomed/issues/1023) is opened so that we continue the discussion about the confusion on  `"slice_axis"` key there
